src/main/src/main/java/org/geoserver/security/impl/DefaultResourceAccessManager.java

```diff
@@ -156,14 +191,105 @@ public boolean canAccess(Authentication user, ResourceInfo resource, AccessMode
+        Filter containsResource = Predicates.equal("layers.resource.id", resource.getId(), MatchAction.ANY);
+        Filter notSingle = Predicates.notEqual("mode", LayerGroupInfo.Mode.SINGLE);
+        Filter global = Predicates.isNull("workspace");
+        Filter inSameWorkspace = Predicates.equal("workspace.name", workspace);
+        Filter wsMatch = Predicates.or(global, inSameWorkspace);
+        Filter groupFilter = Predicates.and(wsMatch, notSingle, containsResource);
+        if(hasContainerGroupNotDenied(nodes, groupFilter)) {
+            return true;
+        }
+        // could not find an authorized layer group that will allow authorization, so not authorized
+        return false;
+    }
+
+    private List<SecureTreeNode> getNodesContainingResource(ResourceInfo resource) {
+        String ws = resource.getStore().getWorkspace().getName();
+        final String resourceId = resource.getId();
+        return getNodesContainingId(ws, resourceId);
+    }
+    
+    private List<SecureTreeNode> getNodesContainingGroup(LayerGroupInfo group) {
+        String ws = group.getWorkspace() != null ? group.getWorkspace().getName() : null;
+        final String groupId = group.getId();
+        return getNodesContainingId(ws, groupId);
+    }
+
+    private List<SecureTreeNode> getNodesContainingId(String ws, final String id) {
+        List<SecureTreeNode> nodes = new ArrayList<>();
+        root.getChildren().forEach((name, node) -> {
+            // global layer group, if contains it gets added, there is no change of name overlap
+            if (node.isContainerLayerGroup()) {
+                if (node.getContainedCatalogIds().contains(id)) {
+                    nodes.add(node);
+                }
+            } else if (ws != null && name.equals(ws)) {
+                List<SecureTreeNode> wsNodes = getSubNodesContainingId(node, ws, id);
+                nodes.addAll(wsNodes);
+            }
+        });
+
+        return nodes;
+    }
+
+    private List<SecureTreeNode> getSubNodesContainingId(SecureTreeNode parent, String workspace, String resourceId) {
+        return parent.getChildren().entrySet().stream().filter(
+                entry -> entry.getValue().isContainerLayerGroup() 
+                && entry.getValue().getContainedCatalogIds().contains(resourceId) 
+                && rawCatalog.getLayerByName(workspace + ":" + entry.getKey()) == null).map(entry -> entry.getValue()).collect(Collectors.toList());
+    }
+
+    private boolean layerGroupContainmentCheckRequired() {
+        // first, is it WMS?
+        Request request = Dispatcher.REQUEST.get();
+        if(request == null) {
+            return false;
+        }
+        
+        // layer groups are used only in WMS
+        final String service = request.getService();
+        boolean isMapService = "WMS".equalsIgnoreCase(service) || "gwc".equalsIgnoreCase(service);
+        if(!isMapService) {
+            return false;
+        }
+        
+        return true;
```

Comment:
The last six lines can be squashed into: return "WMS".equalsIgnoreCase(service) || "gwc".equalsIgnoreCase(service);